Sources/Einstein/Shared/Image/ImageCache.swift

```diff
@@ -0,0 +1,57 @@
+//
+//  ImageCache.swift
+//  
+//
+//  Created by Richard Blanchard on 6/23/19.
+//
+
+import Combine
+import Foundation
+import UIKit
+
+public class ImageCache {
+    
+    public enum Error: Swift.Error {
+        case couldNotConvertData(URL)
+        case general(Swift.Error )
+    }
+    
+    private var imageMap: [URL: UIImage] = [:]
+    private let dispatchQueue = DispatchQueue(from: StringKey(rawValue: "com.ImageCache"))
+    
+    public subscript(key: URL) -> Publishers.Future<UIImage, Error>? {
+        get {
+            let imageForKey = image(for: key)
+    
+            imageForKey
+                .sink(receiveValue: { [weak self, key] image in
+                    self?.dispatchQueue.sync {
+                        self?.imageMap[key] = image
+                    }
+                })
+            
+            return imageForKey
+        }
+    }
+    
+    private func image(for key: URL) -> Publishers.Future<UIImage, Error> {
+        if let cachedImageFuture = self[key] {
```

Comment:
Won't this create an infinite loop? Maybe I'm totally wrong, but.

Inside of the subscript the first thing we do is call `let imageForKey = image(for: key)`. Then this line recalls that subscript method.

Reply:
yeah, just noticed that haha